Cython/Utility/CpdefEnums.pyx

```diff
@@ -86,3 +86,37 @@ else:
 {{if enum_doc is not None}}
 __Pyx_globals["{{name}}"].__doc__ = {{ repr(enum_doc) }}
 {{endif}}
+
+
+#################### EnumTypeToPy ####################
+
+@cname("{{funcname}}")
+cdef {{funcname}}({{name}} c_val):
+    cdef object __pyx_enum
+    # There's a complication here: the Python enum wrapping is only generated
+    # for enums defined in the same module that they're used in. Therefore, if
+    # the enum was cimported from a different module, we try to import it.
+    # If that fails we return an int equivalent as the next best option.
+{{if module_name}}
+    try:
+        from {{module_name}} import {{name}} as __pyx_enum
+    except ImportError:
+        import warnings
+        warnings.warn(
+            f"enum class {{name}} not importable from {{module_name}}. "
+            "You are probably using a cpdef enum declared in a .pxd file that "
+            "does not have a .py  or .pyx file.")
+        return <{{underlying_type}}>c_val
+{{else}}
+    __pyx_enum = {{name}}
+{{endif}}
+    # TODO - Cython only manages to optimize C enums to a switch currently
+    if 0:
+        pass
+{{for item in items}}
+    elif c_val == {{name}}.{{item}}:
+       return __pyx_enum.{{item}}
+{{endfor}}
```

Comment:
Hmm, but the plain item names are C enum values, aren't they? They should just be global names in C (and Cython). Doesn't this get turned into a C `switch()` statement?

Reply:
You're right, but I used the dotted names just to avoid writing different code for C enums and c++ enum classes.

In both cases they could be turned into a C switch, but I think Cython fails to see through the attribute lookup.